[SYCL] Add libsycl, a SYCL RT library implementation project (#1)
sergey-semenov/llvm-project: libsycl/CMakeLists.txt

set(LLVM_SUBPROJECT_TITLE "libsycl")

set(LIBSYCL_SOURCE_DIR  ${CMAKE_CURRENT_SOURCE_DIR})
set(LIBSYCL_BINARY_DIR  ${CMAKE_CURRENT_BINARY_DIR})

set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to conform to")
set(CMAKE_CXX_STANDARD_REQUIRED YES)
set(CMAKE_CXX_EXTENSIONS OFF)

#===============================================================================
# Setup CMake Options
#===============================================================================

option(LIBSYCL_ENABLE_WERROR "Treat all warnings as errors in the libsycl project" OFF)

#===============================================================================
# Configure System
#===============================================================================

set(LIBSYCL_INSTALL_INCLUDE_DIR "${CMAKE_INSTALL_INCLUDEDIR}" CACHE STRING
    "Path where libsycl headers should be installed.")

set(LIBSYCL_SHARED_OUTPUT_NAME "sycl" CACHE STRING "Output name for the shared libsycl runtime library.")

if(LLVM_ENABLE_PER_TARGET_RUNTIME_DIR AND NOT APPLE)
  set(LIBSYCL_TARGET_SUBDIR ${LLVM_DEFAULT_TARGET_TRIPLE})
  if(LIBSYCL_LIBDIR_SUBDIR)
    string(APPEND LIBSYCL_TARGET_SUBDIR /${LIBSYCL_LIBDIR_SUBDIR})
  endif()

>   set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR}/${LIBSYCL_TARGET_SUBDIR})

Reviewer: The `libcxx`, `libcxxabi`, and `libunwind` projects all have this additional line in the matching portions of their `CMakeLists.txt` files. Perhaps it should be present here too? I believe the relevant documentation is [here](https://cmake.org/cmake/help/latest/command/cmake_path.html#normalization). The various `cmake_path` additions were recently added by [commit fc3b67a](https://github.com/llvm/llvm-project/commit/fc3b67a).
```suggestion
  cmake_path(NORMAL_PATH LIBSYCL_TARGET_SUBDIR)
  set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR}/${LIBSYCL_TARGET_SUBDIR})
```

Reviewer 2: seems to be nice to have. Will add this.

      "Path where built libsycl libraries should be installed.")
endif()

set(LIBSYCL_INCLUDE_DIR ${LIBSYCL_SOURCE_DIR}/include)
set(LIBSYCL_BUILD_INCLUDE_DIR ${LLVM_BINARY_DIR}/include)

set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})

# The change in LIBSYCL_MAJOR_VERSION must be accompanied with the same update in
# clang/lib/Driver/CMakeLists.txt.
set(LIBSYCL_MAJOR_VERSION 0)
set(LIBSYCL_MINOR_VERSION 1)
set(LIBSYCL_PATCH_VERSION 0)
set(LIBSYCL_VERSION_STRING "${LIBSYCL_MAJOR_VERSION}.${LIBSYCL_MINOR_VERSION}.${LIBSYCL_PATCH_VERSION}")
set(LIBSYCL_ABI_NAMESPACE "V${LIBSYCL_MAJOR_VERSION}" CACHE STRING
    "The inline ABI namespace used by libsycl. It defaults to Vn where `n` is the current ABI version.")

#===============================================================================
# Setup Compiler Flags
#===============================================================================

# Starting CMake 3.26 this property as ON by default. With 3.20+ we still need to set it.
if(MSVC)
  set_property(GLOBAL PROPERTY USE_FOLDERS ON)

> endif()

Reviewer: What is the reason for restricting the `USE_FOLDERS` property to MSVC? Other projects that set this property (llvm, runtimes, compier-rt) don't (some conditionally set it based on a "standalone" build mode).

Reviewer 2: I believe that this option is widely used for a proper project view in VS IDE. it seems to have not much sense to enable it for linux.
BTW even llvm CMakeLists.txt states that they do it specifically for VS (but still unconditionally).
https://github.com/sergey-semenov/llvm-project/blob/main/llvm/CMakeLists.txt#L378

-----
https://cmake.org/cmake/help/latest/prop_gbl/USE_FOLDERS.html
"Not all CMake generators support recording folder details for targets. The [Xcode](https://cmake.org/cmake/help/latest/generator/Xcode.html#generator:Xcode) and [Visual Studio](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#visual-studio-generators) generators are examples of generators that do. Similarly, not all IDEs support presenting targets using folder hierarchies, even if the CMake generator used provides the necessary information."
https://devblogs.microsoft.com/cppblog/cmake-support-in-visual-studio-targets-view-single-file-compilation-and-cache-generation-settings/

Reviewer 2: Although I do agree that this line is a little strange since this option is important for generator and IDE. Why do we have condition with compiler check? I will remove this condition, it should be fine to set it unconditionally.

Reviewer: Excellent, thanks. I agree Linux users likely wouldn't benefit from this setting, but Apple Xcode users presumably would.

This seems like one of those settings that should be more centrally managed in LLVM, but addressing that would be far out of scope for this PR.

# Enable all warnings by default

> if (MSVC)

Reviewer: Style consistency.
```suggestion
if(MSVC)
```

    string(APPEND LIBSYCL_TARGET_SUBDIR /${LIBSYCL_LIBDIR_SUBDIR})
  endif()
  cmake_path(NORMAL_PATH LIBSYCL_TARGET_SUBDIR)
  set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR}/${LIBSYCL_TARGET_SUBDIR})
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LLVM_LIBDIR_SUFFIX}/${LIBSYCL_TARGET_SUBDIR} CACHE STRING
      "Path where built libsycl libraries should be installed.")
  unset(LIBSYCL_TARGET_SUBDIR)
else()
  if(LLVM_LIBRARY_OUTPUT_INTDIR)
    set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR})
  else()
    set(LIBSYCL_LIBRARY_DIR ${CMAKE_BINARY_DIR}/lib${LIBSYCL_LIBDIR_SUFFIX})
  endif()
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LIBSYCL_LIBDIR_SUFFIX} CACHE STRING
      "Path where built libsycl libraries should be installed.")
endif()

set(LIBSYCL_INCLUDE_DIR include)
set(LIBSYCL_BUILD_INCLUDE_DIR ${LLVM_BINARY_DIR}/${LIBSYCL_INCLUDE_DIR})
set(LIBSYCL_SOURCE_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/include)

set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})

# The change in LIBSYCL_MAJOR_VERSION must be accompanied with the same update in

> # clang/lib/Driver/CMakeLists.txt.

Reviewer: This comment doesn't currently make sense as there is no corresponding code in `clang/lib/Driver/CMakeLists.txt` (yet). I think we should try and solve that issue some other way (if/when we get to the point of needing that additional change in `clang/lib/Driver/CMakeLists.txt`).
```suggestion
```

set(LIBSYCL_MAJOR_VERSION 0)
set(LIBSYCL_MINOR_VERSION 1)
set(LIBSYCL_PATCH_VERSION 0)
set(LIBSYCL_VERSION_STRING "${LIBSYCL_MAJOR_VERSION}.${LIBSYCL_MINOR_VERSION}.${LIBSYCL_PATCH_VERSION}")
set(LIBSYCL_ABI_NAMESPACE "V${LIBSYCL_MAJOR_VERSION}" CACHE STRING
    "The inline ABI namespace used by libsycl. It defaults to Vn where `n` is the current ABI version.")

# Generate headers
configure_file("${LIBSYCL_SOURCE_DIR}/src/version.hpp.in" "${LIBSYCL_BUILD_INCLUDE_DIR}/sycl/version.hpp")

# Install generated headers.
install(FILES
  "${LIBSYCL_BUILD_INCLUDE_DIR}/sycl/version.hpp"
  DESTINATION "${LIBSYCL_INCLUDE_DIR}/sycl"

>   COMPONENT sycl-headers)

Reviewer: I suggest colocating this install command with the other install command further down.

Reviewer 2: I moved it down but didn't join them.

#===============================================================================
# Setup Project
#===============================================================================
cmake_minimum_required(VERSION 3.20.0)

set(LLVM_SUBPROJECT_TITLE "libsycl")

set(LIBSYCL_SOURCE_DIR  ${CMAKE_CURRENT_SOURCE_DIR})
set(LIBSYCL_BINARY_DIR  ${CMAKE_CURRENT_BINARY_DIR})

> set(CMAKE_CXX_STANDARD 17)

Reviewer: For consistency with other LLVM libraries:
```suggestion
set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to conform to")
```

> set(CMAKE_CXX_STANDARD_REQUIRED ON)

Reviewer: For consistency with other LLVM libraries:
```suggestion
set(CMAKE_CXX_STANDARD_REQUIRED YES)
```

  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LLVM_LIBDIR_SUFFIX}/${LIBSYCL_TARGET_SUBDIR} CACHE STRING
      "Path where built libsycl libraries should be installed.")
  unset(LIBSYCL_TARGET_SUBDIR)
else()
  if(LLVM_LIBRARY_OUTPUT_INTDIR)
    set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR})
  else()
    set(LIBSYCL_LIBRARY_DIR ${CMAKE_BINARY_DIR}/lib${LIBSYCL_LIBDIR_SUFFIX})
  endif()
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LIBSYCL_LIBDIR_SUFFIX} CACHE STRING
      "Path where built libsycl libraries should be installed.")
endif()

set(LIBSYCL_INCLUDE_DIR ${LIBSYCL_SOURCE_DIR}/include)
set(LIBSYCL_BUILD_INCLUDE_DIR ${LLVM_BINARY_DIR}/include)

set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})

# The change in LIBSYCL_MAJOR_VERSION must be accompanied with the same update in
# clang/lib/Driver/CMakeLists.txt.
set(LIBSYCL_MAJOR_VERSION 0)
set(LIBSYCL_MINOR_VERSION 1)
set(LIBSYCL_PATCH_VERSION 0)
set(LIBSYCL_VERSION_STRING "${LIBSYCL_MAJOR_VERSION}.${LIBSYCL_MINOR_VERSION}.${LIBSYCL_PATCH_VERSION}")

> set(LIBSYCL_ABI_NAMESPACE "V${LIBSYCL_MAJOR_VERSION}" CACHE STRING "The inline ABI namespace used by libsycl. It defaults to Vn where `n` is the current ABI version.")

Reviewer: Line length.
```suggestion
set(LIBSYCL_ABI_NAMESPACE "V${LIBSYCL_MAJOR_VERSION}" CACHE STRING
    "The inline ABI namespace used by libsycl. It defaults to Vn where `n` is the current ABI version.")
```

  # never throw C++ exceptions.
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /EHsc")

  # Add PDB debug information
  list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
  include(CheckLinkerFlag)
  check_linker_flag(CXX "LINKER:/DEBUG" LINKER_SUPPORTS_DEBUG)
  if(LINKER_SUPPORTS_DEBUG)
    # sccache is not compatible with /Zi flag
    if (CMAKE_CXX_COMPILER_LAUNCHER STREQUAL "sccache")
      # CMake may put /Zi by default
      if(CMAKE_BUILD_TYPE STREQUAL "Debug")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
      endif()

>       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Z7")

Reviewer: ```suggestion
      append("/Z7" CMAKE_CXX_FLAGS)
```

    else()

>       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi")

Reviewer: ```suggestion
      append("/Zi" CMAKE_CXX_FLAGS)
```

    endif()
    add_link_options("LINKER:/DEBUG")

    # Enable unreferenced removal and ICF in Release mode.
    check_linker_flag(CXX "LINKER:/OPT:REF,/OPT:ICF" LINKER_SUPPORTS_OPTS)
    if (LINKER_SUPPORTS_OPTS AND uppercase_CMAKE_BUILD_TYPE STREQUAL "RELEASE")
      add_link_options("LINKER:/OPT:REF" "LINKER:/OPT:ICF")
    endif()
  endif()
endif()

# Enable all warnings by default
if (MSVC)

>   set(CMAKE_CXX_FLAGS "/W4 ${CMAKE_CXX_FLAGS}")

Reviewer: Does `/W4` really need to be inserted at the beginning of the options?
```suggestion
  append("/W4" CMAKE_CXX_FLAGS)
```

Reviewer 2: I haven't found reasons to insert in the beginning, will align with linux path

else ()

>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra")

Reviewer: ```suggestion
  append("-Wall -Wextra" CMAKE_CXX_FLAGS)
```

endif()

if(LIBSYCL_ENABLE_WERROR)
  if(MSVC)

>     set(CMAKE_CXX_FLAGS "/WX ${CMAKE_CXX_FLAGS}")

Reviewer: Does `/WX` really need to be inserted at the beginning of the options?
```suggestion
    append("/WX" CMAKE_CXX_FLAGS)
```

  else()

>     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")

Reviewer: ```suggestion
    append("-Werror" CMAKE_CXX_FLAGS)
```

  endif()
endif()

# This is a workaround to detect changes (add or modify) in subtree which
# are not detected by copy_directory command.

> # TODO: Detect removal in the source tree and reflect in the install directory.

Reviewer: Do we ever intend to follow up on this todo item? If we don't have a plan to do so, I recommend removing it or converting it to a note that just describes the limitation.

      "Path where built libsycl libraries should be installed.")
endif()

set(LIBSYCL_INCLUDE_DIR ${LIBSYCL_SOURCE_DIR}/include)
set(LIBSYCL_BUILD_INCLUDE_DIR ${LLVM_BINARY_DIR}/include)

set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})

# The change in LIBSYCL_MAJOR_VERSION must be accompanied with the same update in
# clang/lib/Driver/CMakeLists.txt.
set(LIBSYCL_MAJOR_VERSION 0)
set(LIBSYCL_MINOR_VERSION 1)
set(LIBSYCL_PATCH_VERSION 0)
set(LIBSYCL_VERSION_STRING "${LIBSYCL_MAJOR_VERSION}.${LIBSYCL_MINOR_VERSION}.${LIBSYCL_PATCH_VERSION}")
set(LIBSYCL_ABI_NAMESPACE "V${LIBSYCL_MAJOR_VERSION}" CACHE STRING
    "The inline ABI namespace used by libsycl. It defaults to Vn where `n` is the current ABI version.")

#===============================================================================
# Setup Compiler Flags
#===============================================================================

if(MSVC)
  set_property(GLOBAL PROPERTY USE_FOLDERS ON)
  # Skip asynchronous C++ exceptions catching and assume "extern C" functions
  # never throw C++ exceptions.

>   append("/EHsc" CMAKE_CXX_FLAGS)

Reviewer: We should validate whether it is necessary to explicitly add `/EHsc` to `CMAKE_CXX_FLAGS`. My understanding is that CMake already passes `/EHsc` by default for MSVC (see https://gitlab.kitware.com/cmake/cmake/blob/v3.20.0/Modules/Platform/Windows-MSVC.cmake#L241).

Reviewer 2: I don't see this behavior to be documented in cmake docs. I think it is not ok to rely on cmake **internal** implementation.
I found where this setting was initially added and I don't understand now why it was needed / how and what exact issue it fixes. So I will remove it for now and add (or not) later when will add support of affected functionality.

commit for history: https://github.com/intel/llvm/commit/c13cb8df84418cb5d682f3bbee89090ebb0d00be

Reviewer: Thank you for the commit reference. *Sigh*, yet another commit with nothing useful in the commit message. Let's please strive to provide descriptive commit messages in commits sent upstream.

  # Add PDB debug information
  list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
  include(CheckLinkerFlag)
  check_linker_flag(CXX "LINKER:/DEBUG" LINKER_SUPPORTS_DEBUG)
  if(LINKER_SUPPORTS_DEBUG)

>     # sccache is not compatible with /Zi flag
>     # https://github.com/mozilla/sccache?tab=readme-ov-file#usage

Reviewer: Merge with suggested comment updates below.
```suggestion
```

    if (CMAKE_CXX_COMPILER_LAUNCHER STREQUAL "sccache")

>       # CMake may put /Zi by default

Reviewer: Thank you for providing the sccache link above, that was helpful.
```suggestion
      # FIXME: CMake 3.24 and earlier will pass the /Zi option by default for compilers
      # that target the MSVC ABI. sccache requires debug information to be embedded
      # in object files. The following code substitutes /Zi with /Z7 to satisfy this
      # requirement. When LLVM migrates to a minimum CMake version of 3.25 or
      # newer, the following code can be removed in favor of the following settings:
      #   set(CMAKE_MSVC_DEBUG_INFORMATION_FORMAT Embedded)
      #   cmake_policy(SET CMP0141 NEW)
      # https://github.com/mozilla/sccache?tab=readme-ov-file#usage
```

      if(CMAKE_BUILD_TYPE STREQUAL "Debug")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
      endif()

>       append("/Z7" CMAKE_CXX_FLAGS)

Reviewer: Unconditionally appending `/Z7` here seems surprising. Should this perhaps be in an `else()` block? I would also expect the same replace strategy as above. Likewise, since the above blocks perform a substitution for both C and C++, we should presumably do likewise here (or not bother with the substitutions for `CMAKE_C_FLAGS_*` above).

Reviewer 2: Good question. Let me to summarize my understanding first:
1) we want to generate PDB unconditionally and we check linker support (LINKER_SUPPORTS_DEBUG).
2) then we want to remove any related flags in CMAKE_CXX_FLAGS_<CONFIG>. I would say that append("/Z7" CMAKE_CXX_FLAGS) should be not needed here. Although it looks like we forgot the 4th option here (MinSizeRel) that should be covered then.

Reviewer: I have to take off early today, so this is the last comment I'll be able to respond to today. I'll try to follow up on any other comments/questions you might have over the weekend (though I have a busy weekend ahead, so it might not be until Monday).

I don't see why generation of debug information by the compiler would be tied to linker support. Do you know the history of why we check for `LINKER:/DEBUG` explicitly?

I suspect the omission of `MinSizeRel` is because the code was copy/pasted from https://github.com/mozilla/sccache?tab=readme-ov-file#usage.

    else()

>       append("/Zi" CMAKE_CXX_FLAGS)

Reviewer: Shouldn't the addition of `/Zi` already be controlled by CMake via the build mode? This seems redundant and oddly located inside a `LINKER_SUPPORTS_DEBUG` conditional block. I think all of this checking for `sccache` and mutation of `CMAKE_C[XX]_FLAGS[_*]` settings should be performed outside of the `LINKER_SUPPORTS_DEBUG` check.

Reviewer 2: I think we want to explicitly specify preferred PDB option here, we don't want to rely on implicit behavior.
I also don't understand why it is not appropriate to have this handling under LINKER_SUPPORTS_DEBUG umbrella. These options may imply /DEBUG usage: 
https://learn.microsoft.com/en-us/cpp/build/reference/z7-zi-zi-debug-information-format?view=msvc-170
"Use of /Zi doesn't affect optimizations. However, /Zi does imply /debug. For more information, see [/DEBUG (Generate Debug Info)](https://learn.microsoft.com/en-us/cpp/build/reference/debug-generate-debug-info?view=msvc-170)."

Reviewer: When  `CMAKE_CXX_COMPILER_LAUNCHER STREQUAL "sccache"` is true, I agree that we should replace `/Zi` with `/Z7` for all build configurations. There are a few issues with the existing code in this respect.

1. [`CMAKE_BUILD_TYPE`](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html#variable:CMAKE_BUILD_TYPE) is only relevant for single configuration generators (e.g., POSIX Makefiles). For multiple configuration generators (e.g., MSVC, Ninja, Xcode), [`CMAKE_CONFIGURATION_TYPES`](https://cmake.org/cmake/help/latest/variable/CMAKE_CONFIGURATION_TYPES.html#variable:CMAKE_CONFIGURATION_TYPES) holds the set of configuration types enabled for the generated build. This makes me question whether the code suggested at https://github.com/mozilla/sccache?tab=readme-ov-file#usage actually works as intended for all generators.
2. As you noted elsewhere, the `MinSizeRel` build type is not handled here (assuming the existing code actually works as intended).
3. The checks for the `CMAKE_BUILD_TYPE` value aren't case insensitive (note the use of `uppercase_CMAKE_BUILD_TYPE` in following code).
4. Regardless of whether a `/Zi` -> `/Z7` replacement occurs, the code appends an (additional) `/Z7`.
5. The code doesn't handle replacement of `/Zi` with `/Z7` in `CMAKE_C_FLAGS` or `CMAKE_CXX_FLAGS`.

With regard to the `LINKER_SUPPORTS_DEBUG` check, it is true that the `/Zi` option implies `/debug`, but that is only relevant when the option is passed to a compiler invocation that will also invoke the linker. That isn't the case here since CMake will perform compilation and linker steps separately.

Finally, I don't understand the motivation for the unconditional addition of `/Zi` when sccache is not being used. This seems like something that should be up to the user that invokes `CMake`; building libsycl shouldn't, I think, require that PDB files be generated.

How would you feel about postponing the sccache and PDB related option settings to a later review? That would allow us more time to update and validate this code without holding up the rest.

Reviewer 2: ok, maybe this PDB generation mostly related to intel/llvm approach when we produce nightly packages (Release) where it is nice to have debug information. 
I will remove this for now.

    endif()
    add_link_options("LINKER:/DEBUG")

    # Enable unreferenced removal and ICF in Release mode.
    check_linker_flag(CXX "LINKER:/OPT:REF,/OPT:ICF" LINKER_SUPPORTS_OPTS)
    if (LINKER_SUPPORTS_OPTS AND uppercase_CMAKE_BUILD_TYPE STREQUAL "RELEASE")
      add_link_options("LINKER:/OPT:REF" "LINKER:/OPT:ICF")

>     endif()

Reviewer: As mentioned in another comment, `CMAKE_BUILD_TYPE` is specific to single configuration generators. I think the right solution for cases like this is to use CMake generator expressions. I don't have much experience with them, but I think this should look something like this.
```suggestion
    # The Microsoft linker implicitly preserves unreferenced symbols and
    # duplicate COMDAT symbols when the `/DEBUG` option is passed.
    # This is not desired for release builds.
    check_linker_flag(CXX "LINKER:/OPT:REF,/OPT:ICF" LINKER_SUPPORTS_OPTS)
    if (LINKER_SUPPORTS_OPTS)
      add_link_options($<$<CONFIG:Release>:"LINKER:/OPT:REF" "LINKER:/OPT:ICF">)
    endif()
```

Should these options be enabled for the `RelWithDebInfo` and `MinSizeRel` build configurations? Since these options are intended to reduce the size of generated binaries, perhaps these options should only be enabled for the `MinSizeRel` build configuration.

Perhaps this is worth postponing to a later PR.

Reviewer 2: removed as part of PDB generation handling

#===============================================================================
# Setup Project
#===============================================================================
cmake_minimum_required(VERSION 3.20.0)

set(LLVM_SUBPROJECT_TITLE "libsycl")

set(LIBSYCL_SOURCE_DIR  ${CMAKE_CURRENT_SOURCE_DIR})
set(LIBSYCL_BINARY_DIR  ${CMAKE_CURRENT_BINARY_DIR})

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

#===============================================================================
# Setup CMake Options
#===============================================================================

option(LIBSYCL_ENABLE_WERROR "Treat all warnings as errors in the libsycl project" OFF)

#===============================================================================
# Configure System
#===============================================================================

set(LIBSYCL_INSTALL_INCLUDE_DIR "${CMAKE_INSTALL_INCLUDEDIR}" CACHE STRING
    "Path where libsycl headers should be installed.")

> set(LIBSYCL_INSTALL_LIBRARY_DIR "${CMAKE_INSTALL_BINDIR}" CACHE STRING

Reviewer 3: It seems this variables is being overwritten on line 38 or 47 below. If so, this does not make sense to have it defined here.

Besides, On Linux `CMAKE_INSTALL_BINDIR` is typically `/usr/bin` if project is installed system wide. This location is no place for .so libraries - that's location for executables, applications and scripts. Dynamic libraries on Linux should effectively go to `CMAKE_INSTALL_LIBDIR`.

Reviewer 2: removed

#===============================================================================
# Setup Project
#===============================================================================
cmake_minimum_required(VERSION 3.20.0)

set(LLVM_SUBPROJECT_TITLE "libsycl")

set(LIBSYCL_SOURCE_DIR  ${CMAKE_CURRENT_SOURCE_DIR})
set(LIBSYCL_BINARY_DIR  ${CMAKE_CURRENT_BINARY_DIR})

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

#===============================================================================
# Setup CMake Options
#===============================================================================

> option(LIBSYCL_ENABLE_WERROR "Treat all warnings as errors in the libsycl project" OFF)

Reviewer: Any reason to not make `-Werror` the default? Is this consistent with other LLVM libraries?

Author: All top-level LLVM projects have it set to off by default.

Reviewer 2: seems to be resolved

> option(LIBSYCL_DISABLE_STL_ASSERTIONS "Disable assertions in STL containers when building libsycl" ON)

Reviewer: A comment would be helpful to explain why assertions in STL containers are disabled by default and why this isn't controlled by `LLVM_ENABLE_ASSERTIONS`.

Reviewer 2: I found that it was disabled by default to stabilize device code with "recent" gcc and used in pair with LLVM_ENABLE_ASSERTIONS https://github.com/intel/llvm/commit/ab69ac8e87c05004241ae123bb5324ebed3d18b1

To not provoke questions from community I will remove it for the current PR and add both - option and comment, later when kernel submission is available

Reviewer 2: removed

option(LIBSYCL_ENABLE_STACK_PRINTING "Enables stack printing on crashes of SYCL applications" OFF)

> option(LIBSYCL_DEBUG_SYMBOLS "Builds SYCL runtime libraries with debug symbols" OFF)

Reviewer: A comment would be helpful to explain why debug symbols are disabled by default and why this isn't controlled by `CMAKE_BUILD_TYPE`.

Reviewer 2: 1) not used in the current PR.
2) I found that it relates to debug feature used in CI to provide developer with call stack in case of failure. https://github.com/intel/llvm/commit/b8d955cf9af395a6bf1f977caff2aa87099dd0da#diff-71d8d95a3500ebd59d2402d9365e329bc41293ed5786ffd8f60b1977997bb6e9

Not relevant to the current PR either, I will remove it.

Reviewer 2: removed

#===============================================================================
# Configure System
#===============================================================================

set(LIBSYCL_INSTALL_INCLUDE_DIR "${CMAKE_INSTALL_INCLUDEDIR}" CACHE STRING
    "Path where libsycl headers should be installed.")

> set(LIBSYCL_INSTALL_RUNTIME_DIR "${CMAKE_INSTALL_BINDIR}" CACHE STRING

Reviewer 3: `LIBSYCL_INSTALL_RUNTIME_DIR` seems to be defined but not used.

Reviewer 2: good catch, LIBSYCL_INSTALL_LIBRARY_DIR seems to be used instead, will update PR

Reviewer 2: updated

  set(LIBSYCL_TARGET_SUBDIR ${LLVM_DEFAULT_TARGET_TRIPLE})
  if(LIBSYCL_LIBDIR_SUBDIR)
    string(APPEND LIBSYCL_TARGET_SUBDIR /${LIBSYCL_LIBDIR_SUBDIR})
  endif()
  set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR}/${LIBSYCL_TARGET_SUBDIR})
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LLVM_LIBDIR_SUFFIX}/${LIBSYCL_TARGET_SUBDIR} CACHE STRING
      "Path where built libsycl libraries should be installed.")
  unset(LIBSYCL_TARGET_SUBDIR)
else()
  if(LLVM_LIBRARY_OUTPUT_INTDIR)
    set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR})
  else()
    set(LIBSYCL_LIBRARY_DIR ${CMAKE_BINARY_DIR}/lib${LIBSYCL_LIBDIR_SUFFIX})
  endif()
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LIBSYCL_LIBDIR_SUFFIX} CACHE STRING
      "Path where built libsycl libraries should be installed.")
endif()

set(LIBSYCL_INCLUDE_DIR ${LIBSYCL_SOURCE_DIR}/include)
set(LIBSYCL_BUILD_INCLUDE_DIR ${LLVM_BINARY_DIR}/include)

set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})

# The change in LIBSYCL_MAJOR_VERSION must be accompanied with the same update in

> # llvm/clang/lib/Driver/CMakeLists.txt.

Reviewer: This path is not correct. Perhaps drop `llvm/` or replace it with `<llvm-project>/`.

      if(CMAKE_BUILD_TYPE STREQUAL "Debug")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
      endif()
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Z7")
    else()
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi")
    endif()
    add_link_options("LINKER:/DEBUG")

    # Enable unreferenced removal and ICF in Release mode.
    check_linker_flag(CXX "LINKER:/OPT:REF,/OPT:ICF" LINKER_SUPPORTS_OPTS)
    if (LINKER_SUPPORTS_OPTS AND uppercase_CMAKE_BUILD_TYPE STREQUAL "RELEASE")
      add_link_options("LINKER:/OPT:REF" "LINKER:/OPT:ICF")
    endif()
  endif()

> endif()

Reviewer: What is the motivation for all of this special debug handling? Is support for `sccache` needed? Other LLVM libraries don't seem to do similar things.

Reviewer 2: sccache is used to reduce build latency on CI https://github.com/intel/llvm/issues/5569 and is used in llvm workflow either https://github.com/llvm/llvm-project/blob/main/.github/workflows/llvm-project-tests.yml#L92

sccache has or at least had a limitation for MSVC /Zi option https://github.com/mozilla/sccache/blob/main/README.md
the issue with /Zi seems to be because full file paths are being embedded in resulting objects, leading to possible caching issues.
https://github.com/intel/llvm/issues/5568
Although I found that MSVC behavior may be fixed in: Visual Studio 2019 version 16.3. So I will check if there is any changes in support now. Although sccache community still keep /Z7 usage as recommended.

Reviewer 2: well, no, it is still not supported, so this WA is needed. llvm subprojects try to use CMAKE_MSVC_DEBUG_INFORMATION_FORMAT feature which is introduced in 3.25.
e.g. https://github.com/llvm/llvm-project/blob/main/libc/fuzzing/__support/CMakeLists.txt#L40

min cmake version is 3.20 so switch to new approach will not cover all supported cmake versions.
So I suggest to leave it as it is. 
Please let me know what you think.

Reviewer: Please add a link to the sccache documentation that recommends use of `/Z7` in the comments.

# Enable all warnings by default
if (MSVC)
  set(CMAKE_CXX_FLAGS "/W4 ${CMAKE_CXX_FLAGS}")
else ()

>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wno-deprecated-declarations")

Reviewer: Are `-Wextra` and `-Wno-deprecated-declarations` really desirable? Other LLVM libraries don't use `Wextra` except in a few specialized cases.

Reviewer 2: partially updated by Sergey in https://github.com/sergey-semenov/llvm-project/pull/1/commits/8a6463fe4fd1015636910f7e0dcdd0f949d58887, 

regarding "-Wextra" I believe - yes, it is desirable. I also see that it is used for the following runtimes in llvm:
./libunwind
./libcxxabi
./libcxx

https://github.com/llvm/llvm-project/blob/main/libcxx/CMakeLists.txt#L827
https://github.com/llvm/llvm-project/blame/main/runtimes/cmake/Modules/WarningFlags.cmake#L15

    string(APPEND LIBSYCL_TARGET_SUBDIR /${LIBSYCL_LIBDIR_SUBDIR})
  endif()
  set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR}/${LIBSYCL_TARGET_SUBDIR})
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LLVM_LIBDIR_SUFFIX}/${LIBSYCL_TARGET_SUBDIR} CACHE STRING
      "Path where built libsycl libraries should be installed.")
  unset(LIBSYCL_TARGET_SUBDIR)
else()
  if(LLVM_LIBRARY_OUTPUT_INTDIR)
    set(LIBSYCL_LIBRARY_DIR ${LLVM_LIBRARY_OUTPUT_INTDIR})
  else()
    set(LIBSYCL_LIBRARY_DIR ${CMAKE_BINARY_DIR}/lib${LIBSYCL_LIBDIR_SUFFIX})
  endif()
  set(LIBSYCL_INSTALL_LIBRARY_DIR lib${LIBSYCL_LIBDIR_SUFFIX} CACHE STRING
      "Path where built libsycl libraries should be installed.")
endif()

set(LIBSYCL_INCLUDE_DIR ${LIBSYCL_SOURCE_DIR}/include)
set(LIBSYCL_BUILD_INCLUDE_DIR ${LLVM_BINARY_DIR}/include)

set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBSYCL_LIBRARY_DIR})

# The change in LIBSYCL_MAJOR_VERSION must be accompanied with the same update in
# llvm/clang/lib/Driver/CMakeLists.txt.
#
# See doc/developer/ABIPolicyGuide.md for the meaning when in the middle of

> # development cycle.

Reviewer 4: That file isn't a part of this PR. I'm not sure if I'd want to see the versioning as part of this first PR or not. Maybe just a TODO for now?

Reviewer 2: agree, will remove

Reviewer 2: I removed mention of ABIPolicyGuide but left versioning itself.

        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
      elseif(CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
      endif()
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Z7")
    else()
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi")
    endif()
    add_link_options("LINKER:/DEBUG")

    # Enable unreferenced removal and ICF in Release mode.
    check_linker_flag(CXX "LINKER:/OPT:REF,/OPT:ICF" LINKER_SUPPORTS_OPTS)
    if (LINKER_SUPPORTS_OPTS AND uppercase_CMAKE_BUILD_TYPE STREQUAL "RELEASE")
      add_link_options("LINKER:/OPT:REF" "LINKER:/OPT:ICF")
    endif()
  endif()
endif()

# Enable all warnings by default
if (MSVC)
  set(CMAKE_CXX_FLAGS "/W4 ${CMAKE_CXX_FLAGS}")
else ()
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wno-deprecated-declarations")
endif()

if(LIBSYCL_ENABLE_WERROR)
  if(MSVC)
    set(CMAKE_CXX_FLAGS "/WX ${CMAKE_CXX_FLAGS}")
    add_definitions(

>       -wd4996 # Suppress 'function': was declared deprecated'

Reviewer 4: Why?

Author: Why indeed. Removed (https://github.com/intel/llvm/pull/17904)

    )
  else()
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
  endif()

> endif()

Reviewer: What is the motivation for suppressing warnings for deprecated declarations? Is this because the SYCL library is required to provide functions that are deprecated in the SYCL specification? If so, can those be suppressed on an individual function basis? I think we would want to get warnings where the implementation is using deprecated declarations from elsewhere.

Author: I'm pretty sure that was the original intent. Ignoring all deprecation warnings isn't something we should do, instead we should only ignore deprecations from SYCL API: https://github.com/intel/llvm/pull/17904

# This is a workaround to detect changes (add or modify) in subtree which
# are not detected by copy_directory command.

> # TODO: detect and process remove header/directory case

Reviewer 4: ```suggestion
# TODO: Detect removal in the source tree and reflect in the install directory.
```

file(GLOB_RECURSE HEADERS_IN_SYCL_DIR CONFIGURE_DEPENDS "${LIBSYCL_INCLUDE_DIR}/sycl/*")

string(REPLACE "${LIBSYCL_INCLUDE_DIR}" "${LIBSYCL_BUILD_INCLUDE_DIR}"
  OUT_HEADERS_IN_SYCL_DIR "${HEADERS_IN_SYCL_DIR}")

# Copy SYCL headers from sources to build directory
add_custom_target(sycl-headers

> DEPENDS ${OUT_HEADERS_IN_SYCL_DIR})

Reviewer: Indent, please.
```suggestion
  DEPENDS ${OUT_HEADERS_IN_SYCL_DIR})
```

add_custom_command(
  OUTPUT  ${OUT_HEADERS_IN_SYCL_DIR}
  DEPENDS ${HEADERS_IN_SYCL_DIR}
  COMMAND ${CMAKE_COMMAND} -E copy_directory ${LIBSYCL_INCLUDE_DIR}/sycl ${LIBSYCL_BUILD_INCLUDE_DIR}/sycl

>   COMMENT "Copying SYCL headers...")

Reviewer 4: https://github.com/llvm/llvm-project/blob/main/libcxx/include/CMakeLists.txt does that differently. Can we justify how our project is different or how our approach is better?

Reviewer 2: I think libcxx uses file list for headers to use copy_if_different tool that copies files only if they were changed. copy_directory_if_different is added in later version, in 3.26. 
We use copy_directory that doesn't detect changes but works faster. At early stages I believe there is no noticeable difference. 

https://cmake.org/cmake/help/v3.26/manual/cmake.1.html#cmdoption-cmake-E-arg-copy
